### PR TITLE
Data Collection Acceptance Tests.

### PR DESCRIPTION
**Pull request by harshjain2**
Implemented of Data Collection Acceptance Tests.

Related to : https://github.com/Microsoft/vstest/issues/309

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs [hunk 1]
```diff
@@ -58,6 +58,8 @@ public override void Initialize()
         {
             try
             {
+                this.ProxyDataCollectionManager?.Initialize();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: When can the `ProxyDataCollectionManager` be null?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncherFactory.cs [hunk 2]
```diff
@@ -21,10 +22,12 @@ internal static class DataCollectionLauncherFactory
...
         /// The <see cref="IDataCollectionLauncher"/>.
         /// </returns>
-        internal static IDataCollectionLauncher GetDataCollectorLauncher(string frameworkVersion)
+        internal static IDataCollectionLauncher GetDataCollectorLauncher(IProcessHelper processHelper)
         {
-            if (frameworkVersion.IndexOf("netstandard", StringComparison.OrdinalIgnoreCase) >= 0
-                || frameworkVersion.IndexOf("netcoreapp", StringComparison.OrdinalIgnoreCase) >= 0)
+            var currentProcessPath = processHelper.GetCurrentProcessFileName();
+
+            if (currentProcessPath.EndsWith("dotnet", StringComparison.OrdinalIgnoreCase)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncherFactory.cs]: Why was this change required? Please add comments to imply intent.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DefaultDataCollectionLauncher.cs [hunk 3]
```diff
@@ -57,13 +58,11 @@ internal DefaultDataCollectionLauncher(IProcessHelper processHelper)
         /// <returns>ProcessId of launched Process. 0 means not launched.</returns>
         public virtual int LaunchDataCollector(IDictionary<string, string> environmentVariables, IList<string> commandLineArguments)
         {
-            var currentWorkingDirectory = Path.GetDirectoryName(typeof(DefaultDataCollectionLauncher).GetTypeInfo().Assembly.Location);
             string dataCollectorProcessPath = null, processWorkingDirectory = null;
+            var currentWorkingDirectory = Path.GetDirectoryName(typeof(DefaultDataCollectionLauncher).GetTypeInfo().Assembly.Location);
+            var currentProcessPath = this.processHelper.GetCurrentProcessFileName();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DefaultDataCollectionLauncher.cs]: Where is this used?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 4]
```diff
@@ -297,7 +297,7 @@ private void InitializeInProcDataCollectors(string runSettings)
             {
                 this.IsInProcDataCollectionEnabled = this.InProcDataCollectors.Any();
             }
-        }
+        }
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: nit: remove

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs [hunk 5]
```diff
@@ -179,11 +178,11 @@ private int VerifyParallelSettingAndCalculateParallelLevel(int sourceCount, stri
...
 
-        private IProxyDataCollectionManager GetDataCollectionManager(Architecture architecture, string settingsXml, string targetFramework)
+        private IProxyDataCollectionManager GetDataCollectionManager(string settingsXml)
         {
             try
             {
-                return new ProxyDataCollectionManager(architecture, settingsXml, targetFramework);
+                return new ProxyDataCollectionManager(settingsXml);
             }
             catch (Exception ex)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: When can this throw?

File: test/Microsoft.TestPlatform.AcceptanceTests/DataCollectionTests.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            dcConfigNode.AppendChild(dataCollectorsNode);
+            var dataCollectorNode = doc.CreateElement(Constants.DataCollectorSettingName);
+            dataCollectorsNode.AppendChild(dataCollectorNode);
+
+            foreach (var kvp in dataCollectionAttributes)
+            {
+                dataCollectorNode.SetAttribute(kvp.Key, kvp.Value);
+            }
+
+            Stream stream = new FileHelper().GetStream(destinationRunsettingsPath, FileMode.Create);
```
Comment (codito) [inline test/Microsoft.TestPlatform.AcceptanceTests/DataCollectionTests.cs]: using ...

File: test/Microsoft.TestPlatform.TestUtilities/IntegrationTestBase.cs [hunk 7]
```diff
@@ -21,8 +21,8 @@ namespace Microsoft.TestPlatform.TestUtilities
     public class IntegrationTestBase
     {
         private const string TestSummaryStatusMessageFormat = "Total tests: {0}. Passed: {1}. Failed: {2}. Skipped: {3}";
-        private string standardTestOutput = string.Empty;
-        private string standardTestError = string.Empty;
+        protected string standardTestOutput = string.Empty;
```
Comment (codito) [inline test/Microsoft.TestPlatform.TestUtilities/IntegrationTestBase.cs]: Why do we need this? There are utility methods that make stdout and stderr available to callers.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncherFactory.cs (hunk 2 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncherFactory.cs]: Do we need test coverage for this?

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs [hunk 8]
```diff
@@ -30,13 +30,13 @@ public void Initialize()
         {
             this.mockDataCollectionRequestSender = new DummyDataCollectionRequestSender();
             this.mockDataCollectionLauncher = new DummyDataCollectionLauncher();
-            this.proxyDataCollectionManager = new ProxyDataCollectionManager(Architecture.AnyCPU, string.Empty, this.mockDataCollectionRequestSender, this.mockDataCollectionLauncher);
+            this.proxyDataCollectionManager = new ProxyDataCollectionManager(string.Empty, this.mockDataCollectionRequestSender, this.mockDataCollectionLauncher);
         }
 
         [TestMethod]
         public void InitializeSocketCommunicationShouldInitializeCommunication()
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs]: Update name of the test?

**Review by codito — CHANGES_REQUESTED** [review]
Few suggestions.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs (hunk 1 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncherFactory.cs (hunk 2 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncherFactory.cs]: Added tests for this class and comment.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DefaultDataCollectionLauncher.cs (hunk 3 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DefaultDataCollectionLauncher.cs]: Removed.

File: test/Microsoft.TestPlatform.AcceptanceTests/DataCollectionTests.cs (hunk 6 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.AcceptanceTests/DataCollectionTests.cs]: Done.

File: test/Microsoft.TestPlatform.TestUtilities/IntegrationTestBase.cs (hunk 7 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.TestUtilities/IntegrationTestBase.cs]: Done.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs (hunk 8 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs (hunk 5 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: This method is no more required. Cleaned up.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs [hunk 9]
```diff
@@ -58,8 +58,20 @@ public override void Initialize()
         {
             try
             {
-                this.ProxyDataCollectionManager?.Initialize();
+                this.ProxyDataCollectionManager.Initialize();
+            }
+            catch (Exception ex)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: Should the runner exit in such a failure?

**Review by codito — APPROVED** [review]
LGTM

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs (hunk 9 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: Let's discuss this please before merge.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs [hunk 10]
```diff
@@ -68,45 +69,40 @@ public override void Initialize()
...
                     // On failure in calling BeforeTestRunStart, call AfterTestRunEnd to end DataCollectionProcess
-                    if (this.ProxyDataCollectionManager != null)
-                    {
-                        this.ProxyDataCollectionManager.AfterTestRunEnd(isCanceled: true, runEventsHandler: this.DataCollectionRunEventsHandler);
-
-                        // Set Proxy Data Collection Manager to null as the communication channel is broken.
-                        this.ProxyDataCollectionManager = null;
-                    }
+                    this.ProxyDataCollectionManager.AfterTestRunEnd(isCanceled: true, runEventsHandler: this.DataCollectionRunEventsHandler);
+                    throw new Exception(Resources.Resources.DataCollectorsInitializationFailed, e);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: This doesn't seem right. It will be caught by the catch in line 97.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs [hunk 11]
```diff
@@ -142,32 +138,33 @@ public override int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEvents
         /// <inheritdoc/>
         public override void Cancel()
         {
-            this.ProxyDataCollectionManager?.AfterTestRunEnd(isCanceled: true, runEventsHandler: this.DataCollectionRunEventsHandler);
+            try
+            {
+                this.ProxyDataCollectionManager.AfterTestRunEnd(isCanceled: true, runEventsHandler: this.DataCollectionRunEventsHandler);
+            }
+            catch (Exception ex)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: Suggest to just log the exception in `finally` and remove `catch`. Don't catch `Exception` to throw again another `Exception`.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyExecutionManagerWithDataCollectionTests.cs [hunk 12]
```diff
@@ -38,36 +42,48 @@ public void TestInit()
...
+        [TestMethod]
+        public void InitializeShouldThrowExceptionIfThrownByDataCollectionManager()
+        {
+            this.mockDataCollectionManager.Setup(x => x.Initialize()).Throws<Exception>();
 
-            mockDataCollectionClient.Verify(dc => dc.BeforeTestRunStart(It.IsAny<bool>(), It.IsAny<bool>(), It.IsAny<ITestMessageEventHandler>()), Times.Once);
+            Assert.ThrowsException<Exception>(() =>
+            {
+                this.proxyExecutionManager.Initialize();
+
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyExecutionManagerWithDataCollectionTests.cs]: nit: remove

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs (hunk 11 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: Same pattern above.

**Comment by codito** [comment]
Please merge after fixing the exceptions.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs (hunk 11 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: **Suggest to just log the exception in finally** I don't think logging exception is possible in finally block without having a catch block. 

Do you mean something like this?

try
            {
                this.ProxyDataCollectionManager.AfterTestRunEnd(isCanceled: true, runEventsHandler: this.DataCollectionRunEventsHandler);
            }
            catch (Exception ex)
            {
                if (EqtTrace.IsErrorEnabled)
                {
                    EqtTrace.Error("ProxyExecutionManagerWithDataCollection: Error occured while communicating with DataCollection Process: {0}", ex.Message);
                }

                throw;
            }
            finally
            {
                base.Cancel();
            }
